src/WebJobs.Extensions.DurableTask/Listener/TaskHttpActivityShim.cs

```diff
@@ -35,20 +38,49 @@ public override string Run(TaskContext context, string input)
 
         public async override Task<string> RunAsync(TaskContext context, string rawInput)
         {
-            HttpRequestMessage requestMessage = await this.ReconstructHttpRequestMessage(rawInput);
-            HttpResponseMessage response = await this.httpClient.SendAsync(requestMessage);
-            DurableHttpResponse durableHttpResponse = await DurableHttpResponse.CreateDurableHttpResponseWithHttpResponseMessage(response);
+            DurableHttpRequest durableHttpRequest = ReconstructDurableHttpRequest(rawInput);
+            HttpRequestMessage requestMessage = await this.ReconstructHttpRequestMessage(durableHttpRequest);
 
-            return JsonConvert.SerializeObject(durableHttpResponse);
+            using (CancellationTokenSource cts = new CancellationTokenSource())
+            {
+                try
+                {
+                    HttpResponseMessage response;
+                    if (durableHttpRequest.Timeout == null)
+                    {
+                        response = await this.httpClient.SendAsync(requestMessage);
+                    }
+                    else
+                    {
+                        cts.CancelAfter(durableHttpRequest.Timeout.Value);
+                        response = await this.httpClient.SendAsync(requestMessage, cts.Token);
+                    }
+
+                    DurableHttpResponse durableHttpResponse = await DurableHttpResponse.CreateDurableHttpResponseWithHttpResponseMessage(response);
+
+                    return JsonConvert.SerializeObject(durableHttpResponse);
+                }
+                catch (OperationCanceledException ex) // when (cts.Token.IsCancellationRequested)
+                {
+                    TimeoutException e = new TimeoutException(ex.Message + $" Reached user specified timeout: {durableHttpRequest.Timeout.Value}.");
+
+                    string details = Utils.SerializeCause(e, this.config.ErrorDataConverter);
+                    throw new TaskFailureException(e.Message, e, details);
+                }
+            }
         }
 
-        private async Task<HttpRequestMessage> ReconstructHttpRequestMessage(string serializedRequest)
+        private static DurableHttpRequest ReconstructDurableHttpRequest(string serializedRequest)
         {
             // DeserializeObject deserializes into a List and then the first element
             // of that list is the DurableHttpRequest
             IList<DurableHttpRequest> input = JsonConvert.DeserializeObject<IList<DurableHttpRequest>>(serializedRequest);
             DurableHttpRequest durableHttpRequest = input.First();
+            return durableHttpRequest;
+        }
 
+        private async Task<HttpRequestMessage> ReconstructHttpRequestMessage(DurableHttpRequest durableHttpRequest)
```

Comment:
nit: This could probably be renamed `ConstructHttpRequestMessage()` or `BuildHttpRequestMessage()` or `ConvertToHttpRequestMessage()`, as we never really had a `HttpRequestMessage` in the first place,